lib/src/kbasesearchengine/events/handler/WorkspaceEventHandler.java

```diff
@@ -644,4 +670,153 @@ private static String toWSRefPath(final List<GUID> objectRefPath) {
+                                    StatusEventType.UNPUBLISH_ALL_VERSIONS).
+                            withNullableAccessGroupID(Integer.decode(accessGrpId)).
+                            withNullableObjectID(objectId).
+                            withNullableVersion(ev.getVersion().get()).
+                            withNullableisPublic(latestIsPublic).build();
+                }
+
+            } catch (IOException ex) {
+                throw handleException(ex);
+            } catch (JsonClientException ex) {
+                throw handleException(ex);
+            }
+        }
+        // update a (UN)DELETE_ALL_VERSIONS event
+        if (eventType.equals(StatusEventType.DELETE_ALL_VERSIONS) ||
+                eventType.equals(StatusEventType.UNDELETE_ALL_VERSIONS )) {
+            Map<String, Object> command;
+
+            // if source data object does not exist then update event as a
+            // DELETE_ALL_VERSIONS event
+            command = new HashMap<>();
+            command.put("command", "listObjects");
+            command.put("params", new ListObjectsParams().
+                    withIds(Arrays.asList(Long.decode(accessGrpId))));
+            try {
+                boolean objidExists = false;
+
+                final List<List> objList = ws.getClient().administer(new UObject(command))
+                        .asClassInstance(List.class);
+
+                for (List obj: objList) {
+                    long id = (long)((Integer)obj.get(0)).intValue();
+                    if (id  == Long.decode(objectId)) {
+                        objidExists = true;
+                        break;
+                    }
+                }
+
+                if (!objidExists) {
+                    return StatusEvent.
+                            getBuilder(ev.getStorageObjectType().get(), ev.getTimestamp(),
+                                    StatusEventType.DELETE_ALL_VERSIONS).
+                            withNullableAccessGroupID(Integer.decode(accessGrpId)).
+                            withNullableObjectID(objectId).build();
+                }
+            } catch (IOException ex) {
+                throw handleException(ex);
+            } catch (JsonClientException ex) {
+                throw handleException(ex);
+            }
+
+            // if source data object is deleted then update event as a DELETE_ALL_VERSIONS event
+            command = new HashMap<>();
+            command.put("command", "listObjects");
+            command.put("params", new ListObjectsParams().
+                                             withShowOnlyDeleted(1L).
+                                             withIds(Arrays.asList(Long.decode(accessGrpId))));
+            final List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,
+                    Map<String,String>>> deletedObjList;
+            try {
+                deletedObjList = ws.getClient().administer(new UObject(command))
```

Comment:
I don't quite understand why the stage is needed. If the object is deleted, it'll be caught above right? It seems like only one of the two stages is needed?

Reply:
My understanding is that an object can be (somehow) permanently deleted and it can also be marked as deleted (which is the normal case). Hence the two checks.

Reply:
I'm still not quite grokking this. The first check looks to see if it shows up in the list objects call.
If it doesn't, the code assumes it's deleted and updates the event.

The second check checks to see if the object shows up in a list of deleted objects - but if the object is deleted, the first check will have already returned, so I don't understand when the 2nd check would get excuted.

Of course what we should do is 1) have error codes for error types like the auth service and 2) have a better way of determining the deletion state of an object (like a getExists() method or something).

Reply:
Good point. I removed the second redundant check. So now if the object is not in the listObjects list, it is either permanently deleted or marked as deleted.